python-ev3dev v2 won't allow negative angles for motors

ev3dev version: 4.14.61-ev3dev-2.2.2-ev3
ev3dev-lang-python version: 
ii  python3-ev3dev                 1.2.0                    all                     Python language bindings for ev3dev
ii  python3-ev3dev2               2.0.0~beta1        all                      Python language bindings for ev3dev

I have been exploring EV3 Python v2 in depth over two weeks and have sent dozens of comments to @wasabifan. He has kindly given lots of feedback, suggesting which comments should be raised as issues here. Based on that feedback I am publishing several issues, of which this is no doubt the most serious.

The rule that angles must be positive for **on_for_rotations** and **on_for_degrees** for single motors and motor pairs is catastrophic, it seems to me! Even these lines generate errors:
```
#!/usr/bin/env python3
from ev3dev2.motor import MoveSteering, OUTPUT_B, OUTPUT_C
steer_pair = MoveSteering(OUTPUT_B, OUTPUT_C)
steer_pair.on_for_degrees(steering=100, speed=40, degrees=685)
```
Error: 
Traceback (most recent call last):
  File "/home/robot/MINE/Basics_3.py", line 9, in <module>
    steer_pair.on_for_degrees(steering=100, speed=40, degrees=685)
  File "/usr/lib/python3/dist-packages/ev3dev2/motor.py", line 1899, in on_for_degrees
    MoveTank.on_for_degrees(self, SpeedNativeUnits(left_speed), SpeedNativeUnits(right_speed), degrees, brake, block)
  File "/usr/lib/python3/dist-packages/ev3dev2/motor.py", line 1807, in on_for_degrees
    self.right_motor._set_position_degrees(right_speed_native_units, right_degrees)
  File "/usr/lib/python3/dist-packages/ev3dev2/motor.py", line 869, in _set_position_degrees
    assert degrees >= 0, "degrees is %s, must be >= 0" % degrees
AssertionError: **degrees is -685.0, must be >= 0**

It seems that any steering command using steering < -50 or steering > 50 will fail since this attempts to turn a wheel backwards by setting the angle negative, which is forbidden.
This also fails, with the same error:
`tank_pair.on_for_degrees(left_speed=30, right_speed=-40, degrees=60)`
AssertionError: **degrees is -7.619047619047619, must be >= 0**

And of course this fails
`mB.on_for_degrees(speed_pct=50, degrees=-60)`
AssertionError: **degrees is -60, must be >= 0**

PLEASE allow negative angles! It will make everything so much simpler and anyone smart enough to be able to program in EV3 Python must surely be able to understand that if both the speed and angle are negative then the result will be a turn in the positive angle direction with a positive speed since that is mathematically equivalent to a turn at a negative speed through a negative angle. There is no contradiction. It's a mathematical fact that (-6) * (-3) = (+6) * (+3). That doesn’t mean that -6 is not negative.

You want Python v2 to behave like EV3-G where possible but EV3-G allows negative angles. A negative power and negative angle cause the motor to turn in the positive direction, no problem.

--- a/ev3dev2/motor.py
+++ b/ev3dev2/motor.py
@@ -1808,9 +1808,9 @@ class MoveTank(MotorSet):
 
         if left_speed_native_units > right_speed_native_units:
             left_degrees = degrees
-            right_degrees = float(right_speed / left_speed_native_units) * degrees
+            right_degrees = abs(float(right_speed_native_units / left_speed_native_units) * degrees)
         else:
-            left_degrees = float(left_speed_native_units / right_speed_native_units) * degrees
+            left_degrees = abs(float(left_speed_native_units / right_speed_native_units) * degrees)
             right_degrees = degrees
 
         # Set all parameters

--- a/ev3dev2/motor.py
+++ b/ev3dev2/motor.py
@@ -1752,12 +1752,12 @@ class MoveTank(MotorSet):
         # v_l = d_l/t, v_r = d_r/t
         # therefore, t = d_l/v_l = d_r/v_r
 
-        # TODO: this is wrong. consider negatives.
-        if left_speed_native_units > right_speed_native_units:
+        # larger speed by magnitude is the "outer" wheel, and rotates the full "rotations"
+        if abs(left_speed_native_units) > abs(right_speed_native_units):
             left_rotations = rotations
-            right_rotations = abs(float(right_speed_native_units / left_speed_native_units)) * rotations
+            right_rotations = abs(right_speed_native_units / left_speed_native_units) * rotations
         else:
-            left_rotations = abs(float(left_speed_native_units / right_speed_native_units)) * rotations
+            left_rotations = abs(left_speed_native_units / right_speed_native_units) * rotations
             right_rotations = rotations
 
         # Set all parameters
@@ -1786,11 +1786,12 @@ class MoveTank(MotorSet):
         """
         (left_speed_native_units, right_speed_native_units) = self._unpack_speeds_to_native_units(left_speed, right_speed)
 
-        if left_speed_native_units > right_speed_native_units:
+        # larger speed by magnitude is the "outer" wheel, and rotates the full "degrees"
+        if abs(left_speed_native_units) > abs(right_speed_native_units):
             left_degrees = degrees
-            right_degrees = abs(float(right_speed_native_units / left_speed_native_units)) * degrees
+            right_degrees = abs(right_speed_native_units / left_speed_native_units) * degrees
         else:
-            left_degrees = abs(float(left_speed_native_units / right_speed_native_units)) * degrees
+            left_degrees = abs(left_speed_native_units / right_speed_native_units) * degrees
             right_degrees = degrees
 
         # Set all parameters

--- a/ev3dev2/motor.py
+++ b/ev3dev2/motor.py
@@ -1746,33 +1746,7 @@ class MoveTank(MotorSet):
         ``rotations`` while the motor on the inside will have its requested
         distance calculated according to the expected turn.
         """
-        (left_speed_native_units, right_speed_native_units) = self._unpack_speeds_to_native_units(left_speed, right_speed)
-
-        # proof of the following distance calculation: consider the circle formed by each wheel's path
-        # v_l = d_l/t, v_r = d_r/t
-        # therefore, t = d_l/v_l = d_r/v_r
-
-        # larger speed by magnitude is the "outer" wheel, and rotates the full "rotations"
-        if abs(left_speed_native_units) > abs(right_speed_native_units):
-            left_rotations = rotations
-            right_rotations = abs(right_speed_native_units / left_speed_native_units) * rotations
-        else:
-            left_rotations = abs(left_speed_native_units / right_speed_native_units) * rotations
-            right_rotations = rotations
-
-        # Set all parameters
-        self.left_motor._set_rel_position_degrees_and_speed_sp(left_rotations * 360, left_speed_native_units)
-        self.left_motor._set_brake(brake)
-        self.right_motor.speed_sp = int(round(right_speed_native_units))
-        self.right_motor._set_rel_position_degrees_and_speed_sp(right_rotations * 360, right_speed_native_units)
-        self.right_motor._set_brake(brake)
-
-        # Start the motors
-        self.left_motor.run_to_rel_pos()
-        self.right_motor.run_to_rel_pos()
-
-        if block:
-            self._block()
+        MoveTank.on_for_degrees(self, left_speed, right_speed, rotations * 360, brake, block)
 
     def on_for_degrees(self, left_speed, right_speed, degrees, brake=True, block=True):
         """
@@ -1784,10 +1758,21 @@ class MoveTank(MotorSet):
         ``degrees`` while the motor on the inside will have its requested
         distance calculated according to the expected turn.
         """
+
+        if left_speed == 0 and right_speed == 0 and degrees != 0:
+            raise ValueError("Can't travel a nonzero distance at zero speed")
+
         (left_speed_native_units, right_speed_native_units) = self._unpack_speeds_to_native_units(left_speed, right_speed)
 
+        # proof of the following distance calculation: consider the circle formed by each wheel's path
+        # v_l = d_l/t, v_r = d_r/t
+        # therefore, t = d_l/v_l = d_r/v_r
+
         # larger speed by magnitude is the "outer" wheel, and rotates the full "degrees"
-        if abs(left_speed_native_units) > abs(right_speed_native_units):
+        if degrees == 0:
+            left_degrees = 0
+            right_degrees = 0
+        elif abs(left_speed_native_units) > abs(right_speed_native_units):
             left_degrees = degrees
             right_degrees = abs(right_speed_native_units / left_speed_native_units) * degrees
         else:

--- a/ev3dev2/motor.py
+++ b/ev3dev2/motor.py
@@ -803,13 +803,13 @@ class Motor(Device):
             self._poll.register(self._state, select.POLLPRI)
 
         while True:
+            if cond(self.state):
+                return True
+            
             self._poll.poll(None if timeout is None else timeout)
 
             if timeout is not None and time.time() >= tic + timeout / 1000:
                 return False
-
-            if cond(self.state):
-                return True
 
     def wait_until_not_moving(self, timeout=None):
         """

--- a/ev3dev2/motor.py
+++ b/ev3dev2/motor.py
@@ -1758,20 +1758,16 @@ class MoveTank(MotorSet):
         ``degrees`` while the motor on the inside will have its requested
         distance calculated according to the expected turn.
         """
-
-        if left_speed == 0 and right_speed == 0 and degrees != 0:
-            raise ValueError("Can't travel a nonzero distance at zero speed")
-
         (left_speed_native_units, right_speed_native_units) = self._unpack_speeds_to_native_units(left_speed, right_speed)
 
         # proof of the following distance calculation: consider the circle formed by each wheel's path
         # v_l = d_l/t, v_r = d_r/t
         # therefore, t = d_l/v_l = d_r/v_r
-
+        
+        if degrees == 0 or (left_speed_native_units == 0 and right_speed_native_units == 0):
+            left_degrees = degrees
+            right_degrees = degrees
         # larger speed by magnitude is the "outer" wheel, and rotates the full "degrees"
-        if degrees == 0:
-            left_degrees = 0
-            right_degrees = 0
         elif abs(left_speed_native_units) > abs(right_speed_native_units):
             left_degrees = degrees
             right_degrees = abs(right_speed_native_units / left_speed_native_units) * degrees

--- a/ev3dev2/motor.py
+++ b/ev3dev2/motor.py
@@ -867,25 +867,15 @@ class Motor(Device):
 
         return speed.to_native_units(self)
 
-    def _set_position_rotations(self, speed, rotations):
+    def _set_rel_position_degrees_and_speed_sp(self, degrees, speed):
+        degrees = degrees if speed >= 0 else -degrees
+        speed = abs(speed)
 
-        # +/- speed is used to control direction, rotations must be positive
-        assert rotations >= 0, "rotations is {}, must be >= 0".format(rotations)
+        position_delta = int(round((degrees * self.count_per_rot)/360))
+        speed_sp = int(round(speed))
 
-        if speed > 0:
-            self.position_sp = self.position + int(round(rotations * self.count_per_rot))
-        else:
-            self.position_sp = self.position - int(round(rotations * self.count_per_rot))
-
-    def _set_position_degrees(self, speed, degrees):
-
-        # +/- speed is used to control direction, degrees must be positive
-        assert degrees >= 0, "degrees is %s, must be >= 0" % degrees
-
-        if speed > 0:
-            self.position_sp = self.position + int((degrees * self.count_per_rot)/360)
-        else:
-            self.position_sp = self.position - int((degrees * self.count_per_rot)/360)
+        self.position_sp = position_delta
+        self.speed_sp = speed_sp
 
     def _set_brake(self, brake):
         if brake:
@@ -900,17 +890,13 @@ class Motor(Device):
         ``speed`` can be a percentage or a :class:`ev3dev2.motor.SpeedValue`
         object, enabling use of other units.
         """
-        speed = self._speed_native_units(speed)
+        if speed is None or rotations is None:
+            raise ValueError("Either speed ({}) or rotations ({}) is None".format(self, speed, rotations))
 
-        if not speed or not rotations:
-            log.warning("({}) Either speed ({}) or rotations ({}) is invalid, motor will not move" .format(self, speed, rotations))
-            self._set_brake(brake)
-            return
-
-        self.speed_sp = int(round(speed))
-        self._set_position_rotations(speed, rotations)
+        speed_sp = self._speed_native_units(speed)
+        self._set_rel_position_degrees_and_speed_sp(rotations * 360, speed_sp)
         self._set_brake(brake)
-        self.run_to_abs_pos()
+        self.run_to_rel_pos()
 
         if block:
             self.wait_until('running', timeout=WAIT_RUNNING_TIMEOUT)
@@ -923,17 +909,13 @@ class Motor(Device):
         ``speed`` can be a percentage or a :class:`ev3dev2.motor.SpeedValue`
         object, enabling use of other units.
         """
-        speed = self._speed_native_units(speed)
+        if speed is None or degrees is None:
+            raise ValueError("Either speed ({}) or degrees ({}) is None".format(self, speed, degrees))
 
-        if not speed or not degrees:
-            log.warning("({}) Either speed ({}) or degrees ({}) is invalid, motor will not move".format(self, speed, degrees))
-            self._set_brake(brake)
-            return
-
-        self.speed_sp = int(round(speed))
-        self._set_position_degrees(speed, degrees)
+        speed_sp = self._speed_native_units(speed)
+        self._set_rel_position_degrees_and_speed_sp(degrees, speed_sp)
         self._set_brake(brake)
-        self.run_to_abs_pos()
+        self.run_to_rel_pos()
 
         if block:
             self.wait_until('running', timeout=WAIT_RUNNING_TIMEOUT)
@@ -1748,9 +1730,6 @@ class MoveTank(MotorSet):
     def _unpack_speeds_to_native_units(self, left_speed, right_speed):
         left_speed = self.left_motor._speed_native_units(left_speed, "left_speed")
         right_speed = self.right_motor._speed_native_units(right_speed, "right_speed")
-
-        assert left_speed or right_speed,\
-            "Either left_speed or right_speed must be non-zero"
         
         return (
             left_speed,
@@ -1772,6 +1751,8 @@ class MoveTank(MotorSet):
         # proof of the following distance calculation: consider the circle formed by each wheel's path
         # v_l = d_l/t, v_r = d_r/t
         # therefore, t = d_l/v_l = d_r/v_r
+
+        # TODO: this is wrong. consider negatives.
         if left_speed_native_units > right_speed_native_units:
             left_rotations = rotations
             right_rotations = abs(float(right_speed_native_units / left_speed_native_units)) * rotations
@@ -1780,16 +1761,15 @@ class MoveTank(MotorSet):
             right_rotations = rotations
 
         # Set all parameters
-        self.left_motor.speed_sp = int(round(left_speed_native_units))
-        self.left_motor._set_position_rotations(left_speed_native_units, left_rotations)
+        self.left_motor._set_rel_position_degrees_and_speed_sp(left_rotations * 360, left_speed_native_units)
         self.left_motor._set_brake(brake)
         self.right_motor.speed_sp = int(round(right_speed_native_units))
-        self.right_motor._set_position_rotations(right_speed_native_units, right_rotations)
+        self.right_motor._set_rel_position_degrees_and_speed_sp(right_rotations * 360, right_speed_native_units)
         self.right_motor._set_brake(brake)
 
         # Start the motors
-        self.left_motor.run_to_abs_pos()
-        self.right_motor.run_to_abs_pos()
+        self.left_motor.run_to_rel_pos()
+        self.right_motor.run_to_rel_pos()
 
         if block:
             self._block()
@@ -1808,22 +1788,20 @@ class MoveTank(MotorSet):
 
         if left_speed_native_units > right_speed_native_units:
             left_degrees = degrees
-            right_degrees = abs(float(right_speed_native_units / left_speed_native_units) * degrees)
+            right_degrees = abs(float(right_speed_native_units / left_speed_native_units)) * degrees
         else:
-            left_degrees = abs(float(left_speed_native_units / right_speed_native_units) * degrees)
+            left_degrees = abs(float(left_speed_native_units / right_speed_native_units)) * degrees
             right_degrees = degrees
 
         # Set all parameters
-        self.left_motor.speed_sp = int(round(left_speed_native_units))
-        self.left_motor._set_position_degrees(left_speed_native_units, left_degrees)
+        self.left_motor._set_rel_position_degrees_and_speed_sp(left_degrees, left_speed_native_units)
         self.left_motor._set_brake(brake)
-        self.right_motor.speed_sp = int(round(right_speed_native_units))
-        self.right_motor._set_position_degrees(right_speed_native_units, right_degrees)
+        self.right_motor._set_rel_position_degrees_and_speed_sp(right_degrees, right_speed_native_units)
         self.right_motor._set_brake(brake)
 
         # Start the motors
-        self.left_motor.run_to_abs_pos()
-        self.right_motor.run_to_abs_pos()
+        self.left_motor.run_to_rel_pos()
+        self.right_motor.run_to_rel_pos()
 
         if block:
             self._block()

--- a/tests/api_tests.py
+++ b/tests/api_tests.py
@@ -198,21 +198,64 @@ class TestAPI(unittest.TestCase):
         with self.assertRaises(ValueError):
             m.on_for_rotations(75, None)
 
-    def test_move_tank(self):
+    def test_move_tank_relative_distance(self):
         clean_arena()
         populate_arena([('large_motor', 0, 'outA'), ('large_motor', 1, 'outB')])
 
         drive = MoveTank(OUTPUT_A, OUTPUT_B)
 
+        # simple case (degrees)
+        drive.on_for_degrees(50, 25, 100)
+        self.assertEqual(drive.left_motor.position_sp, 100)
+        self.assertEqual(drive.left_motor.speed_sp, 0.50 * 1050)
+        self.assertEqual(drive.right_motor.position_sp, 50)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0.25 * 1050, delta=0.5)
+        
+        # simple case (rotations, based on degrees)
         drive.on_for_rotations(50, 25, 10)
-
-        self.assertEqual(drive.left_motor.position, 0)
         self.assertEqual(drive.left_motor.position_sp, 10 * 360)
-        self.assertEqual(drive.left_motor.speed_sp, 1050 / 2)
-
-        self.assertEqual(drive.right_motor.position, 0)
+        self.assertEqual(drive.left_motor.speed_sp, 0.50 * 1050)
         self.assertEqual(drive.right_motor.position_sp, 5 * 360)
-        self.assertAlmostEqual(drive.right_motor.speed_sp, 1050 / 4, delta=0.5)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0.25 * 1050, delta=0.5)
+
+        # negative distance
+        drive.on_for_rotations(50, 25, -10)
+        self.assertEqual(drive.left_motor.position_sp, -10 * 360)
+        self.assertEqual(drive.left_motor.speed_sp, 0.50 * 1050)
+        self.assertEqual(drive.right_motor.position_sp, -5 * 360)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0.25 * 1050, delta=0.5)
+
+        # negative speed
+        drive.on_for_rotations(-50, 25, 10)
+        self.assertEqual(drive.left_motor.position_sp, -10 * 360)
+        self.assertEqual(drive.left_motor.speed_sp, 0.50 * 1050)
+        self.assertEqual(drive.right_motor.position_sp, 5 * 360)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0.25 * 1050, delta=0.5)
+
+        # negative distance and speed
+        drive.on_for_rotations(-50, 25, -10)
+        self.assertEqual(drive.left_motor.position_sp, 10 * 360)
+        self.assertEqual(drive.left_motor.speed_sp, 0.50 * 1050)
+        self.assertEqual(drive.right_motor.position_sp, -5 * 360)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0.25 * 1050, delta=0.5)
+
+        # both speeds zero but nonzero distance
+        with self.assertRaises(ValueError):
+            drive.on_for_rotations(0, 0, 10)
+        
+        # zero distance
+        drive.on_for_rotations(25, 50, 0)
+        self.assertEqual(drive.left_motor.position_sp, 0)
+        self.assertAlmostEqual(drive.left_motor.speed_sp, 0.25 * 1050, delta=0.5)
+        self.assertEqual(drive.right_motor.position_sp, 0)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0.50 * 1050)
+
+        # zero distance and zero speed
+        drive.on_for_rotations(0, 0, 0)
+        self.assertEqual(drive.left_motor.position_sp, 0)
+        self.assertAlmostEqual(drive.left_motor.speed_sp, 0)
+        self.assertEqual(drive.right_motor.position_sp, 0)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0)
     
     def test_tank_units(self):
         clean_arena()

--- a/tests/api_tests.py
+++ b/tests/api_tests.py
@@ -172,6 +172,21 @@ class TestAPI(unittest.TestCase):
         self.assertEqual(drive.right_motor.position, 0)
         self.assertEqual(drive.right_motor.position_sp, 5 * 360)
         self.assertEqual(drive.right_motor.speed_sp, 200)
+    
+    def test_steering_large_value(self):
+        clean_arena()
+        populate_arena([('large_motor', 0, 'outA'), ('large_motor', 1, 'outB')])
+
+        drive = MoveSteering(OUTPUT_A, OUTPUT_B)
+        drive.on_for_rotations(-100, SpeedDPS(400), 10)
+
+        self.assertEqual(drive.left_motor.position, 0)
+        self.assertEqual(drive.left_motor.position_sp, 10 * 360)
+        self.assertEqual(drive.left_motor.speed_sp, -400)
+
+        self.assertEqual(drive.right_motor.position, 0)
+        self.assertEqual(drive.right_motor.position_sp, 10 * 360)
+        self.assertEqual(drive.right_motor.speed_sp, 400)
 
     def test_joystick_units(self):
         clean_arena()

--- a/tests/api_tests.py
+++ b/tests/api_tests.py
@@ -11,7 +11,11 @@ from clean_arena    import clean_arena
 
 import ev3dev2
 from ev3dev2.sensor.lego import InfraredSensor
-from ev3dev2.motor import Motor, MediumMotor, MoveTank, MoveSteering, MoveJoystick, SpeedPercent, SpeedDPM, SpeedDPS, SpeedRPM, SpeedRPS, SpeedNativeUnits, OUTPUT_A, OUTPUT_B
+from ev3dev2.motor import \
+    Motor, MediumMotor, LargeMotor, \
+    MoveTank, MoveSteering, MoveJoystick, \
+    SpeedPercent, SpeedDPM, SpeedDPS, SpeedRPM, SpeedRPS, SpeedNativeUnits, \
+    OUTPUT_A, OUTPUT_B
 
 ev3dev2.Device.DEVICE_ROOT_PATH = os.path.join(FAKE_SYS, 'arena')
 
@@ -128,11 +132,78 @@ class TestAPI(unittest.TestCase):
         m.speed_sp = 500
         self.assertEqual(m.speed_sp, 500)
 
+    def test_motor_on_for_degrees(self):
+        clean_arena()
+        populate_arena([('large_motor', 0, 'outA')])
+
+        m = LargeMotor()
+
+        # simple case
+        m.on_for_degrees(75, 100)
+        self.assertEqual(m.speed_sp, int(round(0.75 * 1050)))
+        self.assertEqual(m.position_sp, 100)
+    
+        # various negative cases; values act like multiplication
+        m.on_for_degrees(-75, 100)
+        self.assertEqual(m.speed_sp, int(round(0.75 * 1050)))
+        self.assertEqual(m.position_sp, -100)
+        
+        m.on_for_degrees(75, -100)
+        self.assertEqual(m.speed_sp, int(round(0.75 * 1050)))
+        self.assertEqual(m.position_sp, -100)
+        
+        m.on_for_degrees(-75, -100)
+        self.assertEqual(m.speed_sp, int(round(0.75 * 1050)))
+        self.assertEqual(m.position_sp, 100)
+
+        # zero speed (on-device, this will return immediately due to reported stall)
+        m.on_for_degrees(0, 100)
+        self.assertEqual(m.speed_sp, 0)
+        self.assertEqual(m.position_sp, 100)
+        
+        # zero distance
+        m.on_for_degrees(75, 0)
+        self.assertEqual(m.speed_sp, int(round(0.75 * 1050)))
+        self.assertEqual(m.position_sp, 0)
+
+        # zero speed and distance
+        m.on_for_degrees(0, 0)
+        self.assertEqual(m.speed_sp, 0)
+        self.assertEqual(m.position_sp, 0)
+
+        # None speed
+        with self.assertRaises(ValueError):
+            m.on_for_degrees(None, 100)
+        
+        # None distance
+        with self.assertRaises(ValueError):
+            m.on_for_degrees(75, None)
+    
+    def test_motor_on_for_rotations(self):
+        clean_arena()
+        populate_arena([('large_motor', 0, 'outA')])
+
+        m = LargeMotor()
+
+        # simple case
+        m.on_for_rotations(75, 5)
+        self.assertEqual(m.speed_sp, int(round(0.75 * 1050)))
+        self.assertEqual(m.position_sp, 5 * 360)
+
+        # None speed
+        with self.assertRaises(ValueError):
+            m.on_for_rotations(None, 5)
+        
+        # None distance
+        with self.assertRaises(ValueError):
+            m.on_for_rotations(75, None)
+
     def test_move_tank(self):
         clean_arena()
         populate_arena([('large_motor', 0, 'outA'), ('large_motor', 1, 'outB')])
 
         drive = MoveTank(OUTPUT_A, OUTPUT_B)
+
         drive.on_for_rotations(50, 25, 10)
 
         self.assertEqual(drive.left_motor.position, 0)
@@ -181,8 +252,8 @@ class TestAPI(unittest.TestCase):
         drive.on_for_rotations(-100, SpeedDPS(400), 10)
 
         self.assertEqual(drive.left_motor.position, 0)
-        self.assertEqual(drive.left_motor.position_sp, 10 * 360)
-        self.assertEqual(drive.left_motor.speed_sp, -400)
+        self.assertEqual(drive.left_motor.position_sp, -10 * 360)
+        self.assertEqual(drive.left_motor.speed_sp, 400)
 
         self.assertEqual(drive.right_motor.position, 0)
         self.assertEqual(drive.right_motor.position_sp, 10 * 360)

--- a/tests/api_tests.py
+++ b/tests/api_tests.py
@@ -240,8 +240,11 @@ class TestAPI(unittest.TestCase):
         self.assertAlmostEqual(drive.right_motor.speed_sp, 0.25 * 1050, delta=0.5)
 
         # both speeds zero but nonzero distance
-        with self.assertRaises(ValueError):
-            drive.on_for_rotations(0, 0, 10)
+        drive.on_for_rotations(0, 0, 10)
+        self.assertEqual(drive.left_motor.position_sp, 10 * 360)
+        self.assertAlmostEqual(drive.left_motor.speed_sp, 0)
+        self.assertEqual(drive.right_motor.position_sp, 10 * 360)
+        self.assertAlmostEqual(drive.right_motor.speed_sp, 0)
         
         # zero distance
         drive.on_for_rotations(25, 50, 0)

